core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java

```diff
@@ -0,0 +1,707 @@
+        maxSize * conf.getFloat(PropertyKey.MASTER_METASTORE_INODE_CACHE_HIGH_WATER_MARK_RATIO));
+    int lowWaterMark = Math.round(
+        maxSize * conf.getFloat(PropertyKey.MASTER_METASTORE_INODE_CACHE_LOW_WATER_MARK_RATIO));
+
+    mFullyCached = true;
+    CacheConfiguration cacheConf = CacheConfiguration.newBuilder()
+        .setMaxSize(maxSize)
+        .setHighWaterMark(highWaterMark)
+        .setLowWaterMark(lowWaterMark)
+        .setEvictBatchSize(conf.getInt(PropertyKey.MASTER_METASTORE_INODE_CACHE_EVICT_BATCH_SIZE))
+        .build();
+    mInodeCache = new InodeCache(cacheConf);
+    mEdgeCache = new EdgeCache(cacheConf);
+    mListingCache = new ListingCache(cacheConf);
+  }
+
+  @Override
+  public Optional<MutableInode<?>> getMutable(long id) {
+    return mInodeCache.get(id);
+  }
+
+  @Override
+  public void remove(Long inodeId) {
+    mInodeCache.remove(inodeId);
+  }
+
+  @Override
+  public void writeInode(MutableInode<?> inode) {
+    mInodeCache.put(inode.getId(), inode);
+  }
+
+  @Override
+  public void writeNewInode(MutableInode<?> inode) {
+    if (inode.isDirectory()) {
+      mListingCache.addEmptyDirectory(inode.getId());
+    }
+    mInodeCache.put(inode.getId(), inode);
+  }
+
+  @Override
+  public void clear() {
+    mInodeCache.clear();
+    mEdgeCache.clear();
+    mBackingStore.clear();
+  }
+
+  @Override
+  public void addChild(long parentId, String childName, Long childId) {
+    mEdgeCache.put(new Edge(parentId, childName), childId);
+  }
+
+  @Override
+  public void removeChild(long parentId, String name) {
+    mEdgeCache.remove(new Edge(parentId, name));
+  }
+
+  @Override
+  public long estimateSize() {
+    return mBackingStore.estimateSize();
+  }
+
+  @Override
+  public Iterable<Long> getChildIds(Long inodeId) {
+    return () -> mListingCache.getChildIds(inodeId).iterator();
+  }
+
+  @Override
+  public Optional<Long> getChildId(Long inodeId, String name) {
+    return mEdgeCache.get(new Edge(inodeId, name));
+  }
+
+  @Override
+  public Optional<Inode> getChild(Long inodeId, String name) {
+    return mEdgeCache.get(new Edge(inodeId, name)).flatMap(this::get);
+  }
+
+  @Override
+  public boolean hasChildren(InodeDirectoryView inode) {
+    Optional<Collection<Long>> cached = mListingCache.getCachedChildIds(inode.getId());
+    if (cached.isPresent()) {
+      return !cached.get().isEmpty();
+    }
+    return !mEdgeCache.getChildIds(inode.getId()).isEmpty() || mBackingStore.hasChildren(inode);
+  }
+
+  /**
+   * Cache for inode metadata.
+   *
+   * The cache supports high concurrency across different inode ids, but requires external
+   * synchronization for operations on the same inode id. All inodes modifications must hold at
+   * least an mLockManager read lock on the modified inode. This allows the cache to flush inodes
+   * asynchronously by acquiring a write lock before serializing the inode.
+   */
+  private class InodeCache extends Cache<Long, MutableInode<?>> {
```

Comment:
They share a lot of fields with the outer class, and there's a circular dependency between EdgeCache and ListingCache. It would be messy to split them up